src/manifolds/Symplectic.jl

```diff
@@ -0,0 +1,889 @@
+``p \in \operatorname{Sp}(2n)``.
+
+The element ``\text{grad}f(p)`` is found as the Riesz representer of the differential
+``\text{D}f(p) \colon T_p\operatorname{Sp}(2n) \rightarrow ℝ`` w.r.t.
+the Riemannian metric inner product at ``p`` [^FioriSimone2011].
+That is, ``\text{grad}f(p) \in T_p\operatorname{Sp}(2n)`` solves the relation
+````math
+    g_p(\text{grad}f(p), X) = \text{D}f(p) \quad\forall\; X \in T_p\operatorname{Sp}(2n).
+````
+
+The default behaviour is to first change the representation of the Euclidean gradient from
+the Euclidean metric to the [`RealSymplecticMetric`](@ref) at ``p``, and then we projecting
+the result onto the correct tangent tangent space ``T_p\operatorname{Sp}(2n, ℝ)``
+w.r.t the Riemannian metric ``g_p`` extended to the entire embedding space.
+
+# Arguments:
+- `extended_metric = true`: If `true`, compute the gradient ``\text{grad}f(p)`` by
+    first changing the representer of the Euclidean gradient of a smooth extension
+    of ``f``, ``∇f(p)``, w.r.t. the [`RealSymplecticMetric`](@ref) at ``p``
+    extended to the entire embedding space, before projecting onto the correct
+    tangent vector space w.r.t. the same extended metric ``g_p``.
+    If `false`, compute the gradient by first projecting ``∇f(p)`` onto the
+    tangent vector space, before changing the representer in the tangent
+    vector space to comply with the [`RealSymplecticMetric`](@ref).
+
+
+[^FioriSimone2011]:
+    > Simone Fiori:
+    > Solving minimal-distance problems over the manifold of real-symplectic matrices,
+    > SIAM Journal on Matrix Analysis and Applications 32(3), pp. 938-968, 2011.
+    > doi [10.1137/100817115](https://doi.org/10.1137/100817115).
+"""
+function gradient(
+    M::Symplectic,
+    f,
+    p,
+    backend::RiemannianProjectionBackend;
+    extended_metric=true,
+)
+    Y = allocate_result(M, gradient, p)
+    return gradient!(M, f, Y, p, backend, extended_metric=extended_metric)
+end
+
+function gradient!(
+    M::Symplectic,
+    f,
+    X,
+    p,
+    backend::RiemannianProjectionBackend;
+    extended_metric=true,
+)
+    _gradient!(f, X, p, backend.diff_backend)
+    if extended_metric
+        change_representer!(
+            MetricManifold(get_embedding(M), ExtendedSymplecticMetric()),
+            X,
+            EuclideanMetric(),
+            p,
+            X,
+        )
+        return project_riemannian!(M, X, p, X)
+    else
+        project!(M, X, p, X)
+        return change_representer!(M, X, EuclideanMetric(), p, X)
+    end
+end
+
+@doc raw"""
+    inner(::Symplectic{n, ℝ}, p, X, Y)
+
+Compute the canonical Riemannian inner product [`RealSymplecticMetric`](@ref)
+````math
+    g_p(X, Y) = \operatorname{tr}((p^{-1}X)^T (p^{-1}Y))
+````
+between the two tangent vectors ``X, Y \in T_p\operatorname{Sp}(2n)``.
+"""
+function inner(M::Symplectic{n,ℝ}, p, X, Y)::eltype(p) where {n}
+    p_star = inv(M, p)
+    return tr((p_star * X)' * (p_star * Y))
```

Comment:
```suggestion
    return dot((p_star * X), (p_star * Y))
```
I didn't check carefully but usually dot(a, b) is the same as tr(a'*b) and the dot version is faster.

Reply:
:+1:

Reply:
Sorry I seem to have mixed up some of the commits, I thought I'd already commited this change, but I'll reintroduce it now at least.